File: Dynamic_Programming/063.Unique-Paths-II/063.Unique-Paths-II.ts
function uniquePathsWithObstacles(obstacleGrid: number[][]): number {
  const m = obstacleGrid.length;
  const n = obstacleGrid[0].length;
  const dp = new Array(m).fill(null).map(() => new Array(n).fill(0));
  dp[0][0] = 1;

  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (obstacleGrid[i][j] === 1) {
        dp[i][j] = 0;
      } else {
        if (i === 0 && j === 0) {
          continue;
        } else if (i === 0) {
          dp[i][j] = dp[i][j - 1];
        } else if (j === 0) {
          dp[i][j] = dp[i - 1][j];
        } else {
          dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
        }
      }
    }
  }
  return dp[m - 1][n - 1];
}

File: Dynamic_Programming/063.Unique-Paths-II/063.Unique-Paths-II.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        m = len(obstacleGrid)
        n = len(obstacleGrid[0])

        if obstacleGrid[0][0] == 1 or obstacleGrid[m - 1][n - 1] == 1:
            return 0

        paths = [[0] * n for i in range(m)]
        for i in range(m):
            for j in range(n):
                if i == 0 and j == 0:
                    paths[i][j] = 1
                elif obstacleGrid[i][j] == 1:
                    paths[i][j] = 0
                elif i == 0:
                    paths[i][j] = paths[i][j - 1]
                elif j == 0:
                    paths[i][j] = paths[i - 1][j]
                else:
                    paths[i][j] = paths[i - 1][j] + paths[i][j - 1]
        return paths[m - 1][n - 1]
        
